### PR TITLE
Allow loading schedule from a file

### PR DESCRIPTION
**Pull request by filipbartek**
Call `vampire --schedule file --schedule_file schedule.txt` to load the schedule from the input file "schedule.txt". Each line of "schedule.txt" is an encoded strategy.

# Known issues
- [x] If the input schedule file contains a line that is not a valid encoding of a strategy, Vampire fails in a dirty manner.

### CODE REVIEW

File: CASC/PortfolioMode.cpp [hunk 1]
```diff
@@ -178,6 +178,10 @@ bool PortfolioMode::performStrategy(Shell::Property* property)
     schedules.push(main);
   }
 
+  if (all_of(schedules.begin(), schedules.end(), [](const Schedule& schedule){ return schedule.isEmpty(); })) {
+    USER_ERROR("The schedule is empty.");
```
Comment (filipbartek) [inline CASC/PortfolioMode.cpp:182]: Otherwise we would end up in a hopeless infinite loop.

File: CASC/Schedules.cpp [hunk 2]
```diff
@@ -16,10 +16,29 @@
...
+#include <fstream>
+
 using namespace Lib;
 using namespace Shell;
 using namespace CASC;
 
+void Schedules::getScheduleFromFile(const vstring& filename, Schedule& quick)
+{
+  if (filename == "") {
+    USER_ERROR("Schedule file was not set.");
```
Comment (filipbartek) [inline CASC/Schedules.cpp:28]: I wonder if the default / invalid value can ever be "<empty>".

File: Shell/Options.cpp [hunk 3]
```diff
@@ -155,10 +156,15 @@ void Options::init()
          "smtcomp",
          "smtcomp_2018",
          "struct_induction"});
-    _schedule.description = "Schedule to be run by the portfolio mode. casc and smtcomp usually point to the most recent schedule in that category. Note that some old schedules may contain option values that are no longer supported - see ignore_missing.";
+    _schedule.description = "Schedule to be run by the portfolio mode. casc and smtcomp usually point to the most recent schedule in that category. file loads the schedule from a file specified in --schedule_file. Note that some old schedules may contain option values that are no longer supported - see ignore_missing.";
     _lookup.insert(&_schedule);
     _schedule.reliesOn(UsingPortfolioTechnology());
 
+    _scheduleFile = StringOptionValue("schedule_file", "", "");
+    _scheduleFile.description = "Path to the input schedule file. Each line contains an encoded strategy. Disabled unless `--schedule file` is set.";
```
Comment (filipbartek) [inline Shell/Options.cpp:164]: The backticks around `--schedule file` follow a convention used in some descriptions but not in the description of `_schedule`.

File: CASC/Schedules.cpp (hunk 2 above)
Comment (MichaelRawson) [reply CASC/Schedules.cpp:28]: I think this is fine. If somebody supplies an empty string somehow they probably don't have a file with an empty name.

**Review by MichaelRawson — APPROVED** [review]
Nice feature! Looks all OK to me. You can get slightly further validating input without much effort if you do something like
```
  while (getline(schedule_file, line)) {
    Options opts;
    try {
      opts.readFromEncodedOptions(line);
      opts.checkGlobalOptionConstraints();
    }
    catch (...) {
      USER_ERROR("bad strat: " + line);
    }
    quick.push(line);
  }
```
but it still doesn't completely validate it as you can have e.g. invalid options.

**Comment by filipbartek** [comment]
> Nice feature! Looks all OK to me. You can get slightly further validating input without much effort if you do something like
> 
> ```
>   while (getline(schedule_file, line)) {
>     Options opts;
>     try {
>       opts.readFromEncodedOptions(line);
>       opts.checkGlobalOptionConstraints();
>     }
>     catch (...) {
>       USER_ERROR("bad strat: " + line);
>     }
>     quick.push(line);
>   }
> ```
Thanks for the suggestion, Michael! I've implemented it in 9f37455.

File: CASC/Schedules.cpp [hunk 4]
```diff
@@ -16,10 +16,45 @@
...
+  if (schedule_file.fail()) {
+    USER_ERROR("Cannot open schedule file: " + filename);
+  }
+  ASS(schedule_file.is_open());
+  vstring line;
+  while (getline(schedule_file, line)) {
+    // Allow structuring the schedule file with empty lines.
+    // Allow documenting the schedule file with line comments.
+    // Interpret lines that start with '%' as comments, following the TPTP convention.
+    if (line == "" or line[0] == '%') {
```
Comment (filipbartek) [inline CASC/Schedules.cpp:43]: > Comments extend from a % character to the end of the line
Source: https://www.tptp.org/TPTP/TR/TPTPTR.shtml

In SMT-LIB, comments start with a `;`.

**Comment by filipbartek** [comment]
@quicquid I've resolved all Michael's comments. I believe this pull request is ready to be merged.

**Comment by MichaelRawson** [comment]
Maybe @quickbeam123? But otherwise I agree!

**Pull request merged**